### PR TITLE
Subclass dnd plus cleanup.

### PR DESCRIPTION
**Pull request by ryan-rushton**

### CODE REVIEW

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 1]
```diff
@@ -154,6 +148,31 @@ export function addItem(
...
+  socketOverrides?: SocketOverrides
+): LoadoutUpdateFunction {
+  const updateFunctions: LoadoutUpdateFunction[] = [];
+  if (item.bucket.hash === BucketHashes.Subclass) {
+    updateFunctions.push(clearSubclass(defs));
+  }
+  updateFunctions.push(replaceEquippedItem(defs, item, Boolean(equip)));
+  updateFunctions.push(addItem(defs, item, equip, socketOverrides));
+  return compose(...updateFunctions);
+}
```
Comment (ryan-rushton) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:174]: Want opinions on this, rather than redefining all the logic for equipping and adding items I have composed a bunch of actions for dropping. I realise this is a bit different to current patterns, which are to extract functionality out into a helper function. Happy to move over to that if people prefer.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 2]
```diff
@@ -194,64 +213,80 @@ export function removeItem(
...
+          (item.equippingLabel && equippingLabel === item.equippingLabel)
+        ) {
+          li.equip = false;
+        }
+      }
+      loadoutItem.equip = true;
+    }
+  }
+}
+
```
Comment (ryan-rushton) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:275]: Extracted out of the `equipItem` action for this so it can be used in a few places.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 1 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:174]: Going to remove this and extract functions like we normally do so I can just get it merged

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 3]
```diff
@@ -52,6 +52,11 @@ import {
  */
 export type LoadoutUpdateFunction = (loadout: Loadout) => Loadout;
 
+/** A helper for composing loadout update functions in the reducer */
```
Comment (bhollis) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:55]: Would be good to say what "composing loadout update functions" means, like why you would use this. I think it just runs each provided update function in sequence?

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 4]
```diff
@@ -154,6 +148,31 @@ export function addItem(
...
+  equip?: boolean,
+  socketOverrides?: SocketOverrides
+): LoadoutUpdateFunction {
+  const updateFunctions: LoadoutUpdateFunction[] = [];
+  if (item.bucket.hash === BucketHashes.Subclass) {
+    updateFunctions.push(clearSubclass(defs));
+  }
+  updateFunctions.push(replaceEquippedItem(defs, item, Boolean(equip)));
+  updateFunctions.push(addItem(defs, item, equip, socketOverrides));
+  return compose(...updateFunctions);
```
Comment (bhollis) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:173]: Would this be easier to read if it were written like this? I'm not sure.

```suggestion
  return (loadout) => {
    if (item.bucket.hash === BucketHashes.Subclass) {
      loadout = clearSubclass(defs)(loadout);
    }
    loadout = replaceEquippedItem(defs, item, Boolean(equip))(loadout);
    loadout = addItem(defs, item, equip, socketOverrides)(loadout);
    return loadout;
  };
```

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 1 above)
Comment (bhollis) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:174]: I don't really mind it (I get what you're doing) but it veers into the "build a data structure to model logic" style of programming that can get hard to follow.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 5]
```diff
@@ -194,64 +213,80 @@ export function removeItem(
   });
 }
 
+function setEquipForItemInLoadout(
```
Comment (bhollis) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:216]: Would be great to have documentation on this

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 6]
```diff
@@ -194,64 +213,80 @@ export function removeItem(
...
+  defs: D1ManifestDefinitions | D2ManifestDefinitions,
+  item: DimItem,
+  loadout: Loadout,
+  isEquipped: boolean
```
Comment (bhollis) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:220]: This is the desired state?

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 7]
```diff
@@ -194,64 +213,80 @@ export function removeItem(
...
+) {
+  // Subclasses and some others are always equipped
+  if (singularBucketHashes.includes(item.bucket.hash)) {
+    return;
+  }
+
+  // TODO: it might be nice if we just assigned a unique ID to every loadout item just for in-memory ops like deleting
+  // We can't just look it up by identity since Immer wraps objects in a proxy and getItemsFromLoadoutItems
+  // changes the socketOverrides, so simply search by unmodified ID and hash.
```
Comment (bhollis) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:229]: Agreed. If we operated on the resolved loadout items this'd also be much easier. Maybe that's something to think about for later.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts [hunk 8]
```diff
@@ -194,64 +213,80 @@ export function removeItem(
...
+
+  if (loadoutItemIndex === -1) {
+    return;
+  }
+  const loadoutItem = loadout.items[loadoutItemIndex];
+
+  if (item.equipment) {
+    if (!isEquipped) {
+      // It's equipped, mark it unequipped
+      loadoutItem.equip = false;
```
Comment (bhollis) [inline src/app/loadout-drawer/loadout-drawer-reducer.ts:244]: Looks like this is mutating the loadout. Should it explicitly operate on `Draft<Loadout>`?

File: src/app/loadout/loadout-edit/LoadoutEditSubclass.m.scss [hunk 9]
```diff
@@ -55,3 +55,13 @@
...
   opacity: 0.3;
 }
+
+.canDrop {
+  filter: invert(20%);
+  background-color: rgba(255, 255, 255, 0.2);
+}
+
+.isOver {
+  filter: invert(40%);
```
Comment (bhollis) [inline src/app/loadout/loadout-edit/LoadoutEditSubclass.m.scss:65]: I'm having trouble imagining what this looks like...

File: src/app/loadout/loadout-edit/useEquipDropTargets.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
+import { DimItem } from 'app/inventory/item-types';
+import { singularBucketHashes } from 'app/loadout-drawer/loadout-utils';
+import { itemCanBeInLoadout } from 'app/utils/item-utils';
+import { DestinyClass } from 'bungie-api-ts/destiny2';
+import { TargetType } from 'dnd-core';
+import { DropTargetHookSpec, useDrop } from 'react-dnd';
+
+/**
+ * Hook to setup dnd drops for loadout editing.
```
Comment (bhollis) [inline src/app/loadout/loadout-edit/useEquipDropTargets.tsx:9]: Would be great to document this - what it is used for, example of how to use it.

File: src/app/loadout/loadout-edit/useEquipDropTargets.tsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        (i.classType === DestinyClass.Unknown || classType === i.classType) &&
+        (type === 'equipped' || !singularBucketHashes.includes(i.bucket.hash)),
+      collect: (monitor) => ({
+        isOver: monitor.isOver() && monitor.canDrop(),
+        canDrop: monitor.canDrop(),
+      }),
+    });
+
+  const [{ isOver: isOverEquipped, canDrop: canDropEquipped }, equippedRef] = useDrop(
+    dropSpec('equipped'),
```
Comment (bhollis) [inline src/app/loadout/loadout-edit/useEquipDropTargets.tsx:34]: Do the dropSpecs need to be memoized?

**Review by bhollis — APPROVED** [review]

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 1 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:174]: Yeah that is why I wanted opinions. I might just extract some stuff and make it more functional.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 4 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:173]: Yeah I think it would, I think I have already convinced myself that we don't need the composition thing but I am not super keen on the `action(...args)(loadout)` that happens when you use actions directly. I might only partially do this as the `addItem` action is the big one I was avoiding splitting up in this PR.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 8 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:244]: I think that is a good idea, hadn't thought about that.

File: src/app/loadout/loadout-edit/useEquipDropTargets.tsx (hunk 11 above)
Comment (ryan-rushton) [reply src/app/loadout/loadout-edit/useEquipDropTargets.tsx:34]: Not sure, will look into it.

File: src/app/loadout/loadout-edit/useEquipDropTargets.tsx (hunk 10 above)
Comment (ryan-rushton) [reply src/app/loadout/loadout-edit/useEquipDropTargets.tsx:9]: Can do

File: src/app/loadout/loadout-edit/LoadoutEditSubclass.m.scss (hunk 9 above)
Comment (ryan-rushton) [reply src/app/loadout/loadout-edit/LoadoutEditSubclass.m.scss:65]: You should see it in the video on the description.

Screenshot 2023-05-04 at 2 02 50 pm

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 5 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:216]: Can do

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 6 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:220]: Yeah from memory it was.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 4 above)
Comment (bhollis) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:173]: Ypu could always make an "applyUpdate" function to make it less weird, though I don't really mind this sort of thing.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 8 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:246]: Done

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 7 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:231]: This was a comment I copy pasted I think haha. I might actually do this in my next PR.

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 6 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:222]: Put some docs in for it

File: src/app/loadout-drawer/loadout-drawer-reducer.ts (hunk 5 above)
Comment (ryan-rushton) [reply src/app/loadout-drawer/loadout-drawer-reducer.ts:218]: Done

File: src/app/loadout/loadout-edit/useEquipDropTargets.tsx (hunk 11 above)
Comment (ryan-rushton) [reply src/app/loadout/loadout-edit/useEquipDropTargets.tsx:34]: No it doesn't, see https://react-dnd.github.io/react-dnd/docs/api/use-drop. Basically the dependency array we are passing in as the second param is used in their own memoization.

Comment (ryan-rushton) [reply src/app/loadout/loadout-edit/useEquipDropTargets.tsx:34]: From my understanding, while this will get recreated it won't actually change the implementation returned by `useDrop` which should be fine.

Comment (ryan-rushton) [reply src/app/loadout/loadout-edit/useEquipDropTargets.tsx:34]: Happy to wrap it in a `useMemo` if you prefer though, no strong opinion.

File: src/app/loadout/loadout-edit/useEquipDropTargets.tsx (hunk 10 above)
Comment (ryan-rushton) [reply src/app/loadout/loadout-edit/useEquipDropTargets.tsx]: Done

**Comment by nev-r** [comment]
> 2\. (should copy abilities, aspects and fragments).

not sure this is actually working?
